### ISSUE TITLE
Consider splitting protobuf plugin into several ones for separating capabilities from conventions

### ISSUE THREAD

**Issue by vlsi** [google/protobuf-gradle-plugin#629]
See https://docs.gradle.org/current/userguide/designing_gradle_plugins.html#capabilities-vs-conventions

It would be nice, if adding protobuf plugin to a java library would automatically configure protoc tasks with a reasonable default protoc version.

In other words:

```diff
 plugins {
     id("java-library")
+     id("com.google.protobuf") version "0.8.19" // it would automatically configure protoc for "main" source set
 }
```

```diff
 plugins {
     id("java-library")
+     id("com.google.protobuf.base") version "0.8.19" // It would skip all default protoc mapping, and it would defer to the user
 }

 dependencies {
     testImplementation("org.junit.jupiter:junit-jupiter:5.9.1")
+.   // The user will be able to fine-tune protobuf configuration
+    protobufCompiler("com.google.protobuf:protoc:3.21.7")
+    protobufPlugin("io.grpc:protoc-gen-grpc-java:1.50.0")
 }
```

Multiple plugins can be shipped within a single jar, and the key point is that `com.google.protobuf.base` would implement basic set of APIs, configurations, etc, and `com.google.protobuf` (or `com.google.protobuf.auto` or whatever you call it) would `apply(plugin = "com.google.protobuf.base")`, then it would configure protoc for the main sourceSet when `java-library` plugin is detected.

**Comment by ejona86** [comment]
> It would be nice, if adding protobuf plugin to a java library would automatically configure protoc tasks with a reasonable default protoc version.

I don't think this is a good idea. The protoc version has implications for the runtime version, and I don't think it appropriate to add the runtime dependency. Using the "default" version is also probably not good, as it will generally be out-of-date.

We are looking to separate parts of the plugin, but in my mind that's to split out some of the current convention, not to add any more.

Is there anything in this request other than specifying default versions?

**Comment by vlsi** [comment]
>Is there anything in this request other than specifying default versions?

Of course. For instance, configure protoc tasks for the main sourceset.
I can easily imagine someone would like to opt-out and add custom wiring, so at least two plugins are needed: the base, and the one with default task wiring.

**Comment by ejona86** [comment]
You might take a look at 0.9.x and see if it does what you're talking about with the main sourceset.``

**Comment by rougsig** [comment]
The main idea here is to divide the plugin into minimum required features (basic) and default features (as it is now). This is useful for advanced Gradle users, more flexibility.

As it explained [here](https://docs.gradle.org/current/userguide/designing_gradle_plugins.html#capabilities-vs-conventions) base version should provide minimal required features. In our case it can be:
* protoc download
* plugins dependency download
* task classes (but not tasks registration)
* etc

**Comment by ejona86** [comment]
> I can easily imagine someone would like to opt-out and add custom wiring

What in particular are _you_ needing/wanting?

> The main idea hereis to divide the plugin into minimum required features (basic) and default features (as it is now).

It isn't clear that's what this issue is about. If that's what this issue is about, then it should state what is going wrong now; what convention is causing harm. It needs that as a problem statement and then the "split the plugin" is a proposed solution.

It sounds like instead the issue is about wanting more functionality, but that functionality would be opinionated so splitting the plugin into more parts is a technical solution to avoid causing problems to users that don't share those opinions. @vlsi, if that is inaccurate, please do correct me.
